### PR TITLE
libuuu: fix race condition in device scan logic

### PR DESCRIPTION
**Pull request by agners**
When using libusb without libudev we encountered a Segmentation fault:
  Thread 1 "uuu" received signal SIGSEGV, Segmentation fault.
  0x0000000000000000 in ?? ()
  (gdb) bt
  #0  0x0000000000000000 in ?? ()
  #1  0x00000000006736b1 in __gthread_detach (__threadid=<optimized out>) at /usr/src/debug/gcc-runtime/9.2.0-r0/x86_64-poky-linux/libstdc++-v3/include/x86_64-poky-linux/bits/gthr-default.h:675
  #2  std::thread::detach (this=this@entry=0x7fffffffdb88) at ../../../../../../../../../../work-shared/gcc-9.2.0-r0/gcc-9.2.0/libstdc++-v3/src/c++11/thread.cc:124
  #3  0x0000000000436d7d in usb_add (dev=<optimized out>) at /usr/include/c++/9.2.0/bits/unique_ptr.h:75
  #4  0x0000000000436f85 in compare_list (old=<optimized out>, nw=<optimized out>) at /usr/src/debug/mfgtools/1.3.102+gitAUTOINC+b078bd087f-r0/git/libuuu/usbhotplug.cpp:181
  #5  compare_list (old=0x0, nw=<optimized out>) at /usr/src/debug/mfgtools/1.3.102+gitAUTOINC+b078bd087f-r0/git/libuuu/usbhotplug.cpp:172
  #6  0x0000000000437042 in polling_usb (bexit=...) at /usr/src/debug/mfgtools/1.3.102+gitAUTOINC+b078bd087f-r0/git/libuuu/usbhotplug.cpp:241
  #7  0x00000000004220b8 in uuu_wait_uuu_finish (deamon=deamon@entry=0, dry=dry@entry=0) at /usr/src/debug/mfgtools/1.3.102+gitAUTOINC+b078bd087f-r0/git/libuuu/cmd.cpp:782
  #8  0x0000000000409024 in main (argc=<optimized out>, argv=<optimized out>) at /usr/src/debug/mfgtools/1.3.102+gitAUTOINC+b078bd087f-r0/git/uuu/uuu.cpp:1057

What seems to fix the issue is wait for the processing thread to
finish (using join() instead of detach()). I guess it is realted to
the libusb device reference which is no longer valid after compare_list
is finished. Joining on the thread seems to work reliably and seems to
have no adversive effects otherwise.

Signed-off-by: Stefan Agner <stefan.agner@toradex.com>

### CODE REVIEW

**Comment by agners** [comment]
It turns out that this is actually a linking problem. We built it statically linked, and it seems that the pthread function with the detach implementation is not properly linked, see also:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52590

That being said, doesn't join make more sense here? What is the point of continuing scanning when we have an actual match we can work on?

**Comment by nxpfrankli** [comment]
> It turns out that this is actually a linking problem. We built it statically linked, and it seems that the pthread function with the detach implementation is not properly linked, see also:
> https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52590
> 
> That being said, doesn't join make more sense here? What is the point of continuing scanning when we have an actual match we can work on?

Because uuu support multiboards program parallel when use -d option,   uuu have to continue monitor usb port plug-in and plug-out.  if you have 10 boards, when finish one, you just remove old board and plug-in new board, needn't rerun uuu.

File: libuuu/usbhotplug.cpp [hunk 1]
```diff
@@ -160,7 +160,7 @@ static int usb_add(libusb_device *dev)
 	if (item)
 	{
 		g_known_device_appeared = 1;
-		std::thread(run_usb_cmds, item, dev).detach();
+		std::thread(run_usb_cmds, item, dev).join();
 	}
```
Comment (nxpfrankli) [inline libuuu/usbhotplug.cpp]: can't join here, which will block scan thread to find new boards.  If you plug two boards at the same, the second one have to wait for first one finish.  Lost parallel download feature.

Comment (agners) [reply libuuu/usbhotplug.cpp]: Ok, I see, wasn't aware of that feature. In that case using `detach()` is of course the right thing to do.

With properly linked static build I actually can use `detach()` successfully. So this can be closed.
